# Build Classification Model

In [15]:
import pandas as pd
cuisines_df = pd.read_csv("../data/cleaned_cuisines.csv")
cuisines_df.head()

,Unnamed: 0,cuisine,almond,angelica,anise,anise_seed,apple,apple_brandy,apricot,armagnac,...,whiskey,white_bread,white_wine,whole_grain_wheat_flour,wine,wood,yam,yeast,yogurt,zucchini
0,0,indian,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,1,indian,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,2,indian,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,3,indian,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,4,indian,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0


In [16]:
cuisines_label_df = cuisines_df['cuisine']
cuisines_label_df.head()

0    indian
1    indian
2    indian
3    indian
4    indian
Name: cuisine, dtype: str

In [17]:
cuisines_features_df = cuisines_df.drop(['Unnamed: 0', 'cuisine'], axis=1)
cuisines_features_df.head()

,almond,angelica,anise,anise_seed,apple,apple_brandy,apricot,armagnac,artemisia,artichoke,...,whiskey,white_bread,white_wine,whole_grain_wheat_flour,wine,wood,yam,yeast,yogurt,zucchini
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0


In [18]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.linear_model import LogisticRegression 
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score, precision_score, confusion_matrix, classification_report
import numpy as np

In [19]:
X_train, X_test, y_train, y_test = train_test_split(cuisines_features_df, cuisines_label_df, test_size=0.2)

In [20]:
C = 10
classifiers = {
    'Linear SVC': SVC(kernel='linear', C=C, random_state=0),
    'KNN classifier': KNeighborsClassifier(C),
    'SVC': SVC(),
    'RFST': RandomForestClassifier(n_estimators=100),
    'ADA': AdaBoostClassifier(n_estimators=100)
}

In [21]:
n_classifiers = len(classifiers)

for index, (name, classifier) in enumerate(classifiers.items()):
    classifier.fit(X_train, np.ravel(y_train))

    y_pred = classifier.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print("Accuracy (train) for %s: %0.1f%%" % (name, accuracy * 100))
    print(classification_report(y_test, y_pred))

Accuracy (train) for Linear SVC: 78.8%
              precision    recall  f1-score   support

     chinese       0.72      0.66      0.69       157
      indian       0.89      0.93      0.91       160
    japanese       0.76      0.71      0.73       167
      korean       0.83      0.77      0.80       168
        thai       0.74      0.87      0.80       147

    accuracy                           0.79       799
   macro avg       0.79      0.79      0.79       799
weighted avg       0.79      0.79      0.79       799

Accuracy (train) for KNN classifier: 73.1%
              precision    recall  f1-score   support

     chinese       0.72      0.64      0.68       157
      indian       0.85      0.84      0.85       160
    japanese       0.61      0.76      0.68       167
      korean       0.91      0.58      0.71       168
        thai       0.67      0.85      0.75       147

    accuracy                           0.73       799
   macro avg       0.75      0.73      0.73      

# Parameter Play Assignment — AdaBoost Classifier

Which classifier you chose- AdaBoost

What your starting hypothesis is- reducing the n_estimators might help increase the accuracy because this would reduce the number of times ada runs (marking down its obsession with mistakes) and there is a chance that it would have a positive impact on the result unless we accidently reduce the n_estimators so much that it doesn't give ada the time and numbers needed to read the patterns, meaning lack in train = disasterous results


## Experiment 1: Baseline (n_estimators=100)

In [24]:
ada_baseline = AdaBoostClassifier(n_estimators=100)
ada_baseline.fit(X_train, np.ravel(y_train))
y_pred = ada_baseline.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Baseline AdaBoost Accuracy: %0.1f%%" % (accuracy * 100))
print(classification_report(y_test, y_pred))

Baseline AdaBoost Accuracy: 67.7%
              precision    recall  f1-score   support

     chinese       0.49      0.77      0.60       157
      indian       0.87      0.82      0.84       160
    japanese       0.72      0.56      0.63       167
      korean       0.73      0.70      0.72       168
        thai       0.71      0.53      0.61       147

    accuracy                           0.68       799
   macro avg       0.70      0.68      0.68       799
weighted avg       0.70      0.68      0.68       799



## Experiment 2: Reduced n_estimators (n_estimators=50)
Hypothesis: Fewer estimators means less obsession over noisy misclassified points, potentially improving accuracy.

In [25]:
ada_exp2 = AdaBoostClassifier(n_estimators=50)
ada_exp2.fit(X_train, np.ravel(y_train))
y_pred = ada_exp2.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("AdaBoost n_estimators=50 Accuracy: %0.1f%%" % (accuracy * 100))
print(classification_report(y_test, y_pred))

AdaBoost n_estimators=50 Accuracy: 65.3%
              precision    recall  f1-score   support

     chinese       0.57      0.62      0.60       157
      indian       0.89      0.78      0.83       160
    japanese       0.51      0.60      0.55       167
      korean       0.70      0.73      0.72       168
        thai       0.67      0.52      0.59       147

    accuracy                           0.65       799
   macro avg       0.67      0.65      0.66       799
weighted avg       0.67      0.65      0.66       799



## Result

Unfortunately, the results were disappointing and the AdaBoost accuracy with n_estimators=50 droped to 65.3% when compared to the original which was 67.7%, where n_estimators=100.

It seems like it was the result of adding not getting adiquate number of simulations or test runs during training that has resulted to this result.

AdaBoost builds each estimator sequentially to correct previous mistakes- with only 50 rounds, it didn't have enough correction cycles to learn the complex patterns in the cuisine data.

## Experiment 3: Increased n_estimators (n_estimators=200)
Hypothesis: More estimators means more correction cycles- but risks over-obsessing on noisy points.

In [29]:
ada_exp3 = AdaBoostClassifier(n_estimators=200)
ada_exp3.fit(X_train, np.ravel(y_train))
y_pred = ada_exp3.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("AdaBoost n_estimators=200 Accuracy: %0.1f%%" % (accuracy * 100))
print(classification_report(y_test, y_pred))

AdaBoost n_estimators=200 Accuracy: 69.5%
              precision    recall  f1-score   support

     chinese       0.65      0.57      0.61       157
      indian       0.87      0.82      0.84       160
    japanese       0.54      0.74      0.62       167
      korean       0.82      0.75      0.79       168
        thai       0.67      0.58      0.62       147

    accuracy                           0.69       799
   macro avg       0.71      0.69      0.70       799
weighted avg       0.71      0.69      0.70       799



## Result of Experiment 3
Suprisingly the accuracy increased to 69.5% with n_estimators=200, and it is higher than both the original 67.7% (n_estimators=100) and Experiment 2: 65.3% (n_estimators=50), thus proving my theory wrong and infact bearing the opposite results, where I assumed that reducing the n_esimators would result in better results because I was assuming that the lower results was the outcome of high number of n_estimators

## Final Verdict
Obsession in AdaBoost can only be seen when the n_estimators are very high (more than 500-1000), otherwise each run only increase the accuracy thus proving that more runs in AdaBoost bear good results until or unless they are too many which is much higher in number (generally above 500-1000).

However, this conclusion is specific to this cuisine dataset- noisier or cleaner dataset may show different sweet spots for n_estimators.

## Experiment 4
Let's try lowering ada's learning rate to 0.5 combined with our best n_estimators=200

In [30]:
ada_exp4 = AdaBoostClassifier(n_estimators=200, learning_rate=0.5)
ada_exp4.fit(X_train, np.ravel(y_train))
y_pred = ada_exp4.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("AdaBoost lr=0.5, n=200 Accuracy: %0.1f%%" % (accuracy * 100))
print(classification_report(y_test, y_pred))

AdaBoost lr=0.5, n=200 Accuracy: 68.2%
              precision    recall  f1-score   support

     chinese       0.60      0.64      0.62       157
      indian       0.89      0.80      0.84       160
    japanese       0.56      0.69      0.62       167
      korean       0.79      0.68      0.73       168
        thai       0.64      0.58      0.61       147

    accuracy                           0.68       799
   macro avg       0.69      0.68      0.68       799
weighted avg       0.70      0.68      0.69       799



## Result Experiment 4
Reducing the learning rate has resulted in a negative impact on the accuracy which before was 69.5% in Experiment3 where the learning rate was normal/standard (i.e. 1) while it got low 68.2% here is Experiment 4 where learning rate was reduced to half of it normal rate to 0.5 here.

## Conclusion
While reducing the learning rate increases stability for AdaBoost, it at the same time impacts the speed of adaboost as such increases the need for more runs to get a better result than the one running at normal rate. Meaning you have to sacrifice speed for more stability but at the same time would need to run more cycles (n_estimators), to get better result

## Experiment 5: 
Compensating lower learning_rate with more estimators (learning_rate=0.5, n_estimators=400)

Hypothesis: If lower learning rate needs more correction cycles to compensate, doubling n_estimators to 400 should recover or improve accuracy.

In [31]:
ada_exp5 = AdaBoostClassifier(n_estimators=400, learning_rate=0.5)
ada_exp5.fit(X_train, np.ravel(y_train))
y_pred = ada_exp5.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("AdaBoost lr=0.5, n=400 Accuracy: %0.1f%%" % (accuracy * 100))
print(classification_report(y_test, y_pred))

AdaBoost lr=0.5, n=400 Accuracy: 68.3%
              precision    recall  f1-score   support

     chinese       0.59      0.59      0.59       157
      indian       0.88      0.80      0.84       160
    japanese       0.55      0.73      0.63       167
      korean       0.84      0.70      0.76       168
        thai       0.64      0.59      0.61       147

    accuracy                           0.68       799
   macro avg       0.70      0.68      0.69       799
weighted avg       0.70      0.68      0.69       799



## Conclusion Experiment 5
Contradictory to my hypothesis, that with a lower learning rate, more runs would (double of that we ran at the normal rate) would bring us better or at least the same result, the results did show improvement but only by 0.1%, which doesn't justify going through all the trouble of adjusting the rate and increasing the number of run (also using more computing) power to worse result than when we ran it in experiment 3 at normal rate with 200 runs.

## Final Overall Conclusion
I conclude that on the basis of this experiment, referring to this specific dataset, sometimes just using the default values (like the learning rate) bring better results like Experiment 3 whereas when we go more in depth and tweek each minor thing the result might be worse and not worth the time and effort both on your part and the machines part.

The best performing configuration was Experiment3- n_estimators=200 with default learning_rate=1.0, achieving 69.5% accuracy.